src/build.cc

```diff
@@ -696,6 +688,20 @@ bool Builder::StartEdge(Edge* edge, string* err) {
       return false;
   }
 
+  // Find the most recent mtime of any (existing) non-order-only input
+  Node* most_recent_input = NULL;
+  for (vector<Node*>::iterator i = edge->inputs_.begin();
+       i != edge->inputs_.end() - edge->order_only_deps_; ++i) {
+    if (!(*i)->Stat(disk_interface_, err))
```

Comment:
For reference, this being `Stat` rather than `StatIfNecessary` is the crux of the performance impact of this approach.  It is needed to account for cases in which an input is modified after the build plan was created but before an edge runs.

Reply:
@bradking I had another thought this morning. What if, instead of trying to be intelligent about tracking the most recent input mtime for an output, we simply _touch_ the output file just prior to starting the edge command, and remember the mtime of the file at that point (and record it when processing the successful command later)? If the output doesn't exist yet, we can simply delete it immediately after grabbing the mtime. We'd probably only have to do this for the first output of the edge (instead of attempting to do it for all outputs).

If the command fails, we won't record any new output mtime in the log file, so subsequent runs would still use the last-successful recorded mtime to compare against current inputs.

I think this solves the performance issue (touch/stat of one file per edge vs a stat of all input files per edge) as well as the hole this PR currently has related to inputs not known to the edge when the command is started.

Reply:
Interesting idea.  Do we even need to touch the actual output?  We could touch another file, or even try to just compute the current time ourselves.

Reply:
I started with computing the time ourselves, but determined it's bad to compare filesystem mtimes to system times. This at least keeps ninja comparing filesystem times to each other.

We could touch another file, but I was concerned that a destination file on one filesystem might have different time stamping granularity (or times, in the case of NFS) than whatever temporary file we decided to use.

Reply:
It feels risky to touch the actual output in case the build is interrupted.

Another approach is to touch a `.ninja_tmp` file in the same directory as the first output, get its mtime, and then remove it.

Reply:
Yep, that's exactly what I'm trying in a separate branch right now.

Reply:
This is working. I'm pretty excited about this alternative, it's much less intrusive than this change, is more performant, and I *think* is correct even more of the time.

Reply:
Great!  That approach should probably be in a new PR.